tooling/miniutil/src/build/terminator.rs

```diff
@@ -1,5 +1,243 @@
 use crate::build::*;
 
+impl FunctionBuilder {
+    fn finish_block(&mut self) {
+        let Some(terminator) = self.terminator else {
+            panic!("Cannot finish block as block has no terminator")
+        };
+        let bb = BasicBlock { statements: self.statements, terminator };
+        self.blocks.try_insert(self.block_name, bb).unwrap();
+    }
+
+    fn new_block(&mut self, name: u32) {
+        self.block_name = BbName(Name::from_internal(name));
+        self.statements = Default::default();
+        self.terminator = None;
+    }
+
+    /// create new block without branching
+    fn prepare_next_block(&mut self) {
+        self.new_block(self.next_block);
+        self.next_block += 1;
+    }
+
+    // terminators with 0 following blocks
+    pub fn exit(&mut self) {
+        self.terminator = Some(exit());
+        self.finish_block();
+    }
+
+    pub fn unreachable(&mut self) {
+        self.terminator = Some(unreachable());
+        self.finish_block();
+    }
+
+    pub fn goto(&mut self, dest: BbName) {
+        self.terminator = Some(Terminator::Goto(dest));
+        self.finish_block();
+    }
+
+    pub fn return_(&mut self) {
+        self.terminator = Some(Terminator::Return);
+        self.finish_block();
+    }
+
+    // terminators with 0 or 1 following blocks
+    pub fn call(&mut self, f: FnName, args: &[ArgumentExpr], ret: PlaceExpr, can_return: Ret) {
+        let next_block = match can_return {
+            Ret::No => None,
+            Ret::Yes => Some(BbName(Name::from_internal(self.next_block))),
```

Comment:
Instead of directly fiddling with `next_block`, what about having a `declare_block` function that just returns a fresh block name, and never accessing `self.next_block` outside that function? Then it is less likely that you will accidentally use a block name twice.

Below then instead of `prepare_next_block()` it would be something like `self.set_current_block(next_block)` where `next_block` is what got returned by `declare_block`.